packages/smooth_app/lib/pages/navigator/app_navigator.dart

```diff
@@ -0,0 +1,332 @@
+
+  // Indicates whether [_initAppLanguage] was already called
+  bool isInitialized = false;
+
+  void _initAppLanguage(BuildContext context) {
+    final UserPreferences userPreferences = context.read<UserPreferences>();
+    ProductQuery.setLanguage(context, userPreferences);
+    context.read<ProductPreferences>().refresh();
+
+    // The migration requires the language to be set in the app!
+    context.read<SmoothAppDataImporter>().startMigrationAsync();
+  }
+
+  /// Extract the barcode from a path only if the route have at least 8 digits
+  /// in the second part (we don't care about extra elements)
+  /// Some examples:
+  /// - produit/156164894948
+  /// - product/3017620422003/nutella-ferrero
+  String? _extractProductBarcode(String path) {
+    if (path.isEmpty) {
+      return null;
+    }
+
+    final List<String> pathParams = path.split('/').sublist(1);
+
+    if (pathParams.length > 1) {
+      final String barcode = pathParams[1];
+
+      // Ensure we only have digits and at least 8 characters
+      if (int.tryParse(barcode) != null && barcode.length >= 8) {
+        return barcode;
+      }
+    }
+
+    return null;
+  }
+
+  bool _isAnInternalRoute(String path) {
+    if (path == HOME_PAGE) {
+      return true;
+    } else {
+      return path.startsWith('/_');
+    }
+  }
+
+  //region Onboarding
+
+  bool _isOnboardingComplete(BuildContext context) {
+    return _getCurrentOnboardingPage(context).isOnboardingComplete();
+  }
+
+  Widget _findLastOnboardingPage(BuildContext context) {
+    return _getCurrentOnboardingPage(context).getPageWidget(context);
+  }
+
+  OnboardingPage _getCurrentOnboardingPage(BuildContext context) {
+    final UserPreferences userPreferences = context.read<UserPreferences>();
+
+    final OnboardingPage lastVisitedOnboardingPage =
+        userPreferences.lastVisitedOnboardingPage;
+    return lastVisitedOnboardingPage;
+  }
+
+  //endregion Onboarding
+
+  /// Internal routes
+  /// To differentiate external routes (eg: /product/12345678), we prefix all
+  /// internal routes with an underscore
+  static const String HOME_PAGE = '/';
+  static const String PRODUCT_DETAILS_PAGE = '_product';
+  static const String PRODUCT_LOADER_PAGE = '_product_loader';
+  static const String PRODUCT_CREATOR_PAGE = '_product_creator';
+  static const String PREFERENCES_PAGE = '_preferences';
+  static const String SEARCH_PAGE = '_search';
+  static const String EXTERNAL_PAGE = '_external';
+}
+
+/// A list of internal routes to use with [AppNavigator]
+// TODO(g123k): Improve this with sealed classes (Dart 3 is required)
+// ignore_for_file: non_constant_identifier_names
+class AppRoutes {
+  AppRoutes._();
+
+  // Home page (or walkthrough during the onboarding)
+  static const String HOME = _SmoothGoRouter.HOME_PAGE;
+
+  // Product details (a [Product] is mandatory in the extra)
+  static String PRODUCT(String barcode) =>
+      '/${_SmoothGoRouter.PRODUCT_DETAILS_PAGE}/$barcode';
+
```

Comment:
```suggestion
  final String path;

  // Home page (or walkthrough during the onboarding)
  AppRoutes.home() : path = _SmoothGoRouter.HOME_PAGE;

  // Product details (a [Product] is mandatory in the extra)
  AppRoutes.product(String barcode) : path = 
      '/${_SmoothGoRouter.PRODUCT_DETAILS_PAGE}/$barcode';
```

Would look better that way, I think.

Reply:
That would be better, but it means that each entity contains an attribute for nothing.
And what should we do if you we want no one, but two elements?

Reply:
> That would be better, but it means that each entity contains an attribute for nothing.

The attribute being the `String` that you return, right? I don't see it as a problem, I see it as OOP.

> And what should we do if you we want no one, but two elements?

Then with my OOP solution we can do it, like `AppRoutes.product(String barcode, Color color)`.

Reply:
In your second example, you would still have only the path and merge the two data in the construct you mean?